File: Java/Q24SwapNodesInPairs.java
/*
@b-knd (jingru) on 02 August 2022 10:17:00
*/

class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }
        
        //set head of the list to be the second node (this will be returned in the end)
        ListNode dummyHead = head.next;
        ListNode curr = head;
        ListNode prev = new ListNode();
        
        //change next link in while loop
        while(curr != null && curr.next != null){
            ListNode temp = curr.next;
            prev.next = temp;
            curr.next = curr.next.next;
            temp.next = curr;
            prev = curr;
            curr = curr.next;
        }
        return dummyHead;
    }
}

//Runtime: 0 ms, faster than 100.00% of Java online submissions for Swap Nodes in Pairs.
//Memory Usage: 42.6 MB, less than 5.73% of Java online submissions for Swap Nodes in Pairs.

File: Python/Q24SwapNodesInPairs.py
# @b-knd (jingru) on 07 August 2022 16:11:00

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head
        
        dummyHead = head.next
        curr = head
        prev = ListNode(0)
    
        while curr != None and curr.next != None:
            prev.next = curr.next
            temp = curr.next
            curr.next = curr.next.next
            temp.next = curr
            prev = curr
            curr = curr.next
            
        return dummyHead
            
#Runtime: 36 ms, faster than 84.86% of Python3 online submissions for Swap Nodes in Pairs.
#Memory Usage: 14 MB, less than 18.57% of Python3 online submissions for Swap Nodes in Pairs.
        
        
        
    
